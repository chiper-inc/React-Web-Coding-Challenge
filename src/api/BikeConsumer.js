import api from "./BikeResource";

const Consumer = ({

    getAll: (handle) => api.getAll()
        .then(data => {
            handle({ isResult: "success", data });
        }).catch(error => {
            console.error(error);
            handle({ isResult: "error", error });
        }),

    update: (item, handle) => api.put(item)
        .then((data) => {
            handle({ isResult: "success", item: data });
        }).catch(error => {
            console.error(error);
            handle({ isResult: "error", error });
        }),

    add: (item, handle) => api.post({ description: item, id: null, isActive: true })
        .then((data) => {
            handle({ isResult: "success", item: data });
        }).catch(error => {
            console.error(error);
            handle({ isResult: "error", error });
        }),

    delete: (item, handle) => api.delete(item)
        .then((data) => {
            handle({ isResult: "success", item: data });
        }).catch(error => {
            console.error(error);
            handle({ isResult: "error", error });
        }),
});

export default Consumer;


