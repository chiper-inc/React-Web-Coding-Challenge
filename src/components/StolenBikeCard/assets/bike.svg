<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->
<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://web.resource.org/cc/"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   width="410"
   height="240"
   id="svg2161"
   sodipodi:version="0.32"
   inkscape:version="0.45.1"
   version="1.0"
   sodipodi:docbase="/home/phsi/Desktop/clIPART"
   sodipodi:docname="Bicycle-Silhouette.svg"
   inkscape:output_extension="org.inkscape.output.svg.inkscape">
  <defs
     id="defs2163" />
  <sodipodi:namedview
     id="base"
     pagecolor="#ffffff"
     bordercolor="#666666"
     borderopacity="1.0"
     gridtolerance="10000"
     guidetolerance="10"
     objecttolerance="10"
     inkscape:pageopacity="0.0"
     inkscape:pageshadow="2"
     inkscape:zoom="1.0227179"
     inkscape:cx="207.52431"
     inkscape:cy="106.86492"
     inkscape:document-units="px"
     inkscape:current-layer="layer1"
     height="240px"
     width="410px"
     inkscape:window-width="1280"
     inkscape:window-height="949"
     inkscape:window-x="0"
     inkscape:window-y="25" />
  <metadata
     id="metadata2166">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     inkscape:label="Layer 1"
     inkscape:groupmode="layer"
     id="layer1"
     transform="translate(-183.18751,-412.51844)">
    <path
       style="fill:#000000;fill-opacity:1;stroke:none;stroke-opacity:1"
       d="M -95.40625,-321.46875 C -108.00625,-321.46876 -118.21875,-311.24675 -118.21875,-298.71875 C -118.21875,-295.91074 -117.72675,-293.18575 -116.71875,-290.59375 L -140.96875,-230.625 C -151.84075,-237.24901 -164.3185,-240.856 -177.0625,-241 L -177.125,-237.96875 C -177.413,-237.96875 -177.765,-237.96875 -178.125,-237.96875 C -215.92499,-237.96875 -246.53125,-207.3625 -246.53125,-169.5625 C -246.53125,-131.76251 -215.925,-101.15625 -178.125,-101.15625 C -156.23699,-101.15625 -135.65975,-111.538 -122.84375,-129.25 L -120.375,-127.4375 C -111.447,-139.6775 -106.625,-154.3705 -106.625,-169.5625 C -106.625,-191.73851 -116.932,-212.69225 -134.5,-226.15625 L -130.40625,-236.4375 L -55.375,-171.21875 C -56.815,-168.41075 -57.53125,-165.168 -57.53125,-162 C -57.53125,-150.192 -48.01725,-140.625 -36.28125,-140.625 C -35.417251,-140.62501 -34.70775,-140.678 -33.84375,-140.75 L -29.4375,-126.21875 L -34.125,-126.21875 L -34.125,-118.59375 L -14.6875,-118.59375 L -14.6875,-126.21875 L -21.46875,-126.21875 L -26.65625,-143 C -20.203966,-146.29783 -15.881138,-152.52631 -15.0625,-159.6875 L -2.21875,-160.5625 C 2.3172501,-126.5065 31.32175,-101.15625 65.59375,-101.15625 C 101.73775,-101.15625 131.60225,-129.25 133.90625,-165.25 L 136.875,-165.03125 C 136.94701,-166.61524 137.09375,-168.0505 137.09375,-169.5625 C 137.09375,-208.99761 105.08351,-240.96618 65.65625,-241 C 65.64725,-241 65.62975,-241 65.59375,-241 C 54.505748,-241 43.6235,-238.468 33.6875,-233.5 L 14.125,-267.28125 L 21.375,-283.625 L 41.1875,-292.625 C 43.563499,-293.70501 45.0625,-296.06425 45.0625,-298.65625 C 45.062502,-302.32826 42.1095,-305.34375 38.4375,-305.34375 C 24.973499,-305.77574 9.427,-306.0625 -6.125,-306.0625 C -6.9890001,-306.06249 -7.708,-306.0625 -8.5,-306.0625 C -10.876,-306.06249 -10.9375,-306.0625 -10.9375,-306.0625 C -13.6735,-306.06249 -15.84375,-303.9235 -15.84375,-301.1875 C -15.84375,-299.45949 -14.7715,-297.792 -13.1875,-297 C -4.3314999,-292.10401 4.3845,-287.86625 13.3125,-283.90625 L 6.125,-268 L -117.71875,-268 L -108.4375,-290.875 C -114.5575,-300.95501 -107.21425,-313.9375 -95.40625,-313.9375 C -94.182248,-313.9375 -92.88025,-313.788 -91.65625,-313.5 L -91.59375,-313.5 L -78.1875,-309.96875 L -74.53125,-309.09375 L -72.65625,-316.4375 L -76.3125,-317.375 L -89.78125,-320.84375 C -91.581253,-321.27576 -93.46225,-321.46875 -95.40625,-321.46875 z M -120.75,-260.4375 L 2.71875,-260.4375 L -31.8125,-182.8125 C -33.2525,-183.1005 -34.76925,-183.25 -36.28125,-183.25 C -37.073251,-183.25001 -38.0205,-183.16575 -38.8125,-183.09375 L -43.125,-197 L -37.15625,-197 L -37.15625,-204.5625 L -56.59375,-204.5625 L -56.59375,-197 L -51.125,-197 L -46.15625,-180.78125 C -47.884249,-179.91725 -49.52875,-178.77425 -50.96875,-177.40625 L -127.375,-243.9375 L -120.75,-260.4375 z M 10.21875,-258.78125 L 26.9375,-229.75 C 10.305499,-218.95 -1.294,-201.74975 -4.75,-182.09375 C -5.5420002,-177.91775 -5.90625,-173.8105 -5.90625,-169.5625 C -5.9062502,-168.9865 -5.90625,-168.398 -5.90625,-167.75 L -15.6875,-167.25 C -16.9835,-172.434 -20.30775,-176.995 -24.84375,-179.875 L 9.34375,-256.8125 L 10.21875,-258.78125 z M -178.125,-226.875 C -166.749,-226.875 -155.66299,-223.48275 -146.375,-217.21875 L -159.90625,-183.6875 C -163.14625,-177.2075 -169.69675,-173.09375 -176.96875,-173.09375 C -177.25675,-173.09375 -177.6305,-173.09375 -178.0625,-173.09375 L -178.125,-173.09375 C -180.141,-173.09375 -181.87501,-171.43175 -181.875,-169.34375 C -181.875,-167.25575 -180.1505,-165.59376 -178.0625,-165.59375 C -166.9025,-165.16175 -156.6135,-171.65375 -152.4375,-182.09375 L -152.4375,-181.9375 L -140.03125,-212.34375 C -127.79125,-201.54375 -120.8125,-185.9065 -120.8125,-169.5625 C -120.8125,-137.9545 -146.44499,-112.3125 -178.125,-112.3125 C -209.661,-112.3125 -235.375,-137.9545 -235.375,-169.5625 C -235.375,-201.1705 -209.66099,-226.875 -178.125,-226.875 z M 65.53125,-226.875 C 97.21225,-226.875 122.84375,-201.16099 122.84375,-169.625 C 122.84375,-137.945 97.211254,-112.3125 65.53125,-112.3125 C 37.23525,-112.3125 13.09175,-133.12925 8.84375,-161.28125 L 59.125,-164.59375 C 60.709,-162.64975 63.083254,-161.4375 65.53125,-161.4375 C 70.06725,-161.4375 73.718754,-165.08899 73.71875,-169.625 C 73.71875,-174.017 70.204502,-177.63726 65.8125,-177.78125 L 40.75,-221.1875 C 48.454,-224.9315 56.963254,-226.875 65.53125,-226.875 z M 34.125,-217.4375 L 58.96875,-174.46875 L 57.8125,-172.09375 L 8.34375,-168.78125 C 8.27175,-169.06925 8.2812498,-169.33699 8.28125,-169.625 C 8.28125,-188.849 18.068998,-206.92549 34.125,-217.4375 z "
       id="path5707"
       transform="matrix(1.0426849,0,0,1.0439716,445.24192,753.12269)" />
  </g>
</svg>
