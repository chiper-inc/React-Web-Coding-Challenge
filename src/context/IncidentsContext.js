import React, { useState } from "react";

const Context = React.createContext({});

export function IncidentsContextProvider({ children }) {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  const [incidents, setIncidents] = useState([]);
  const [currentDate, setCurrentDate] = useState("");

  return (
    <Context.Provider
      value={{
        loading,
        error,
        incidents,
        currentDate,
        setLoading,
        setError,
        setIncidents,
        setCurrentDate,
      }}
    >
      {children}
    </Context.Provider>
  );
}

export default Context;
