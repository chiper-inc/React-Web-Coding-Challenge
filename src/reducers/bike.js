export default function reducer(state, action) {

  if (action.key) {
    return state.setIn(action.key, action.payload)
  }
  switch (action.type) {
    case "FILTER":
    
      return state.setIn(['bikes', 'filter'],
        state.getIn(['bikes', 'all'])
          .filter(b => b.title.toLowerCase().includes(state.getIn(['search', 'title']).toLowerCase()))
          .filter(b => b.date_stolen === null || b.date_stolen > state.getIn(['search', 'from']))
          .filter(b => b.date_stolen === null || b.date_stolen < state.getIn(['search', 'to']))
      )
    case "CLEAR":
      return state.setIn(['search', 'title'], '')
        .setIn(['search', 'from'], 0)
        .setIn(['search', 'to'], 999999999999999)
    default:
      return state;
  }
}
