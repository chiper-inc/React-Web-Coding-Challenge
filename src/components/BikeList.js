import React, { useContext, useEffect, useState } from "react";
import Pagination from "react-js-pagination";
import BikeConsumer from "../api/BikeConsumer";
import { Bike } from "../context/bike";



const BikeList = () => {
  const { state, dispatch } = useContext(Bike);
  const [isResult, setIsResult] = useState(null);


  function handlePageChange(pageNumber) {
    dispatch({ key: ['pagination', 'page'], payload: pageNumber });
    dispatch({
      key: ['pagination', 'bikes'], payload: state.getIn(['bikes', 'filter'])
        .slice(pageNumber * 10 - 10, pageNumber * 10 - 1)
    });
  }


  useEffect(() => {
    BikeConsumer.getAll((status) => {
      setIsResult(status.isResult);
      if (status.isResult === 'success') {
       
        dispatch({ key: ['bikes', 'all'], payload: status.data.bikes });
        dispatch({ key: ['bikes', 'filter'], payload: status.data.bikes });
        dispatch({ key: ["pagination", 'bikes'], payload: status.data.bikes.slice(0, 9) });
      }
    })
  }, [state.bikes]);

  useEffect(() => {
    dispatch({
      key: ['pagination', 'bikes'], payload: state.getIn(['bikes', 'filter'])
        .slice(0, 9)
    });
  }, [state.getIn(['search'])]);

  return <>{
    {
      success: <List state={state} handlePageChange={handlePageChange} />,
      error: <div className="alert alert-dark">Error in the request</div>,
      null: <div>Loading</div>
    }[isResult]
  }</>;
}


const List = ({ state, handlePageChange }) => <div className="row">
  <div className="col-md-12">
    <div className="row">
      <div className="col-md-12">
        <br />
        <span className="float-right">{`Total ${state.getIn(['bikes', 'filter']).length}`}</span>
      </div>
    </div>
    <div className="row">
     
      <div className="col-md-12">
        <ul className="list-unstyled" data-testid="all-list">
          {
            state.getIn(['pagination', 'bikes'])
              .map(t => (
                <li key={t.id} className="list-group-item media" >
                  <div className="card mb-3">
                    <div className="row no-gutters">
                      <div className="col-md-4">
                        <img className="card-img" src={t.thumb} alt=""></img>
                      </div>
                      <div className="col-md-8">
                        <div className="card-body">
                          <a href={t.url}>
                            <h5 className="card-title">{t.title}</h5>
                          </a>
                          <p className="card-text">{t.description}</p>
                          <p className="card-text">{`${new Date(t.date_stolen).toUTCString()}, ${t.stolen_location}`}</p>
                          <p className="card-text"><small className="text-muted">Reported {t.year}</small></p>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              ))}
        </ul>
      </div>
    </div>
    <br />
    <div className="row justify-content-center">
      <Pagination
        itemClass="page-item"
        linkClass="page-link"
        activePage={state.getIn(['pagination', 'page'])}
        itemsCountPerPage={10}
        totalItemsCount={state.getIn(['bikes', 'filter']).length || 0}
        pageRangeDisplayed={5}
        onChange={(page) => handlePageChange(page)}
      />
    </div>
  </div>
</div>;

export default BikeList;
