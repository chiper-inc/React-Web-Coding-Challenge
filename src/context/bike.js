import { fromJS } from 'immutable';
import React, { useContext, useReducer } from "react";
import BikeForm from "../components/BikeForm";
import BikeHeader from "../components/BikeHeader";
import BikeList from "../components/BikeList";
import reducers from "../reducers/bike";




// Initial Data
export const Bike = React.createContext(fromJS({
  bikes: {
    all: [],
    filter: []
  },
  search: {
    title: '',
    from: 0,
    to: 999999999999999,
  },
  pagination: {
    page: 1,
    bikes: [],
  }
}));

export default () => {
  const globalStore = useContext(Bike);
  const [state, dispatch] = useReducer(reducers, globalStore);
  return (
    <Bike.Provider value={{ state, dispatch }}>
      <BikeHeader />
      <BikeForm />
      <BikeList />
    </Bike.Provider>
  );
};

