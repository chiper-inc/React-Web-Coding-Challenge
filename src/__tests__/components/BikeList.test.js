import { fromJS } from "immutable";
import React from "react";
import { render } from 'react-testing-library';
import BikeConsumer from "../../api/BikeConsumer";
import BikeList from "../../components/BikeList";
import { Bike } from "../../context/bike";

test("<BikeList /> #display", () => {

  const bikes = [
    { title: "bike 1", id: 1 },
    { title: "bike 2", id: 2 },
    { title: "bike 3", id: 3 }
  ];
  const dispatch = jest.fn();//mock validar el dispatch

  //dependencia que consume la API
  jest.spyOn(BikeConsumer, "getAll");
  BikeConsumer.getAll = jest.fn(cb => cb({ data: { bikes }, isResult: 'success' }));

  //renderiza el contenedor
  const { container, getByTestId } = render(
    <Bike.Provider value={{
      state: fromJS({
        pagination: { bikes, page: 1 },
        search: { title: '', from: 0, to: 99999999999999 },
      }).setIn(['bikes', 'filter'], []), dispatch
    }}>
      <BikeList />
    </Bike.Provider>
  );
  expect(container.getElementsByTagName("li").length).toBe(7);
  //caputar el html de test ID y lo guarda
  expect(getByTestId('all-list').innerHTML).toMatchSnapshot();

  expect(BikeConsumer.getAll).toHaveBeenCalledTimes(1);
  expect(dispatch).toHaveBeenCalledWith({
    "key": [
      "bikes",
      "filter",
    ],
    "payload": [
      {
        "id": 1,
        "title": "bike 1",
      },
      {
        "id": 2,
        "title": "bike 2",
      },
      {
        "id": 3,
        "title": "bike 3",
      },
    ],
  });
});






