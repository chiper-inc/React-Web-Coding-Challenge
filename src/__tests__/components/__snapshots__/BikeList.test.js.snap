// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BikeList /> #display 1`] = `"<li class=\\"list-group-item media\\"><div class=\\"card mb-3\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4\\"><img class=\\"card-img\\" alt=\\"\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><a><h5 class=\\"card-title\\"></h5></a><p class=\\"card-text\\"></p><p class=\\"card-text\\">Invalid Date, undefined</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Reported </small></p></div></div></div></div></li><li class=\\"list-group-item media\\"><div class=\\"card mb-3\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4\\"><img class=\\"card-img\\" alt=\\"\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><a><h5 class=\\"card-title\\"></h5></a><p class=\\"card-text\\"></p><p class=\\"card-text\\">Invalid Date, undefined</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Reported </small></p></div></div></div></div></li><li class=\\"list-group-item media\\"><div class=\\"card mb-3\\"><div class=\\"row no-gutters\\"><div class=\\"col-md-4\\"><img class=\\"card-img\\" alt=\\"\\"></div><div class=\\"col-md-8\\"><div class=\\"card-body\\"><a><h5 class=\\"card-title\\"></h5></a><p class=\\"card-text\\"></p><p class=\\"card-text\\">Invalid Date, undefined</p><p class=\\"card-text\\"><small class=\\"text-muted\\">Reported </small></p></div></div></div></div></li>"`;
